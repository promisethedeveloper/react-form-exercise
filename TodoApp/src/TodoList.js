import React, { useState } from "react";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";

const TodoList = () => {
	const [todos, setTodos] = useState([]);

	// add a new todo
	const create = (newTodo) => {
		setTodos((todos) => [...todos, newTodo]);
	};

	// update a todo
	const update = (id, updatedTask) => {
		setTodos((todos) =>
			todos.map((todo) =>
				todo.id === id ? { ...todo, task: updatedTask } : todo
			)
		);
	};

	// remove a todo
	const remove = (id) => {
		setTodos((todos) => todos.filter((todo) => todo.id !== id));
	};

	const todoComponents = todos.map((todo) => (
		<Todo
			key={todo.id}
			id={todo.id}
			task={todo.task}
			update={update}
			remove={remove}
		/>
	));

	return (
		<div>
			<NewTodoForm addTodo={create} />
			<ul>{todoComponents}</ul>
		</div>
	);
};

export default TodoList;
