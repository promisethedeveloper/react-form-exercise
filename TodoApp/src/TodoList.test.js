import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

const addTodo = (todoList, task = "Practice coding") => {
	const taskInput = todoList.getByLabelText("Task:");
	fireEvent.change(taskInput, { target: { value: task } });
	const submitButton = todoList.getByText("Add Todo");
	fireEvent.click(submitButton);
};

it("should render without crashing", () => {
	render(<TodoList />);
});

it("should match snapshot", () => {
	const { asFragment } = render(<TodoList />);
	expect(asFragment()).toMatchSnapshot();
});

it("should add a todo", () => {
	const list = render(<TodoList />);
	addTodo(list);

	expect(list.getByLabelText("Task:")).toHaveValue("");
	expect(list.getByText("Practice coding")).toBeInTheDocument();
	expect(list.getByText("Edit")).toBeInTheDocument();
	expect(list.getByText("X")).toBeInTheDocument();
});

it("should edit a todo", () => {
	const list = render(<TodoList />);
	addTodo(list);

	fireEvent.click(list.getByText("Edit"));
	const editInput = list.getByDisplayValue("Practice coding");
	fireEvent.change(editInput, { target: { value: "Read a book" } });
	fireEvent.click(list.getByText("Update!"));

	// expect only edited todo to appear
	expect(list.getByText("Read a book")).toBeInTheDocument();
	expect(list.queryByText("Practice coding")).not.toBeInTheDocument();
});

it("should delete a todo", () => {
	const list = render(<TodoList />);
	addTodo(list);

	fireEvent.click(list.getByText("X"));

	// expect todo to be deleted
	expect(list.queryByText("Practice coding")).not.toBeInTheDocument();
});
