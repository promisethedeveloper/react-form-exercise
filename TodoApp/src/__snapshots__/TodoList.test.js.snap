// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <h1>
        Task
      </h1>
      <form>
        <label
          for="task"
        >
          Task:
        </label>
        <input
          id="task"
          name="task"
          type="text"
          value=""
        />
        <button>
          Add Todo
        </button>
      </form>
    </div>
    <ul />
  </div>
</DocumentFragment>
`;
